src/main.rs

```diff
@@ -546,27 +623,55 @@ extern "C" fn primary_init(hart_id: u64, fdt_addr: u64) -> CpuParams {
     HOST_VM.call_once(|| host);
 
     // Return launch parameters for this CPU.
-    CpuParams {
+    Ok(CpuParams {
         satp: PerCpu::this_cpu().page_table().satp(),
-    }
+    })
 }
 
-/// Start salus on the bootstrap CPU. This is called once we switch to the hypervisor page-table.
+/// Rust Entry point for the bootstrap CPU. Calls into init to allow
+/// clean error collection and presentation
+///
+/// Note: this runs with a 1:1 map of physical to virtual mapping.
 #[cfg(not(test))]
 #[no_mangle]
-extern "C" fn primary_main() {
+extern "C" fn primary_init(hart_id: u64, fdt_addr: u64) -> CpuParams {
+    match _primary_init(hart_id, fdt_addr) {
+        Err(e) => {
+            // Safe as nothing can fail before console is setup
+            println!("Error during initialization on bootstrap CPU: {}", e);
+            // Safe as if shutdown fails it triggers a wfi loop (via abort())
+            poweroff();
```

Comment:
I like the idea of controlling what happens on boot error in a better way. I personally think Debug print would just do fine.

Reply:
Took me a while to understand what you mean here. But I think you mean there shouldn't be a `Display` implementation for the top level Error or any of the other errors. I guess that's a matter of personal preference - I think that the way nested errors are presented with chains of round brackets aren't the nicest and it can be less obvious what the error is than the human written message. But it would reduce the compiled size by not having the strings and also reduce the amount of code.

Reply:
Sure, definitely a matter of personal preference and not important for the sake of this patch. Regarding the message, it's a panic, will require interpretation. I think the amount of code and data required for the string is definitely negligible.